server/src/main/java/org/elasticsearch/search/sort/RawBucketedSort.java

```diff
@@ -0,0 +1,340 @@
+        if (values().size() < requiredSize) {
+            grow(requiredSize);
+        }
+        int next = getNextGatherOffset(rootIndex);
+        assert 0 <= next && next < bucketSize
+            : "Expected next to be in the range of valid buckets [0 <= " + next + " < " + bucketSize + "]";
+        long index = next + rootIndex;
+        swap(0, index);
+        if (next == 0) {
+            heapMode.set(bucket);
+            heapify(rootIndex);
+        } else {
+            setNextGatherOffset(rootIndex, next - 1);
+        }
+    }
+
+    /**
+     * Superclass for implementations of {@linkplain RawBucketedSort} for {@code long} keys.
+     */
+    public static class ForLongs extends RawBucketedSort<Long> {
+        private LongArray values;
+
+        @SuppressWarnings("this-escape")
+        public ForLongs(BigArrays bigArrays, SortOrder sortOrder, int bucketSize) {
+            super(bigArrays, sortOrder, bucketSize);
+            boolean success = false;
+            try {
+                values = bigArrays.newLongArray(1, false);
+                success = true;
+            } finally {
+                if (success == false) {
+                    close();
+                }
+            }
+        }
+
+        @Override
+        protected final BigArray values() {
+            return values;
+        }
+
+        @Override
+        protected final void growValues(long minSize) {
+            values = bigArrays.grow(values, minSize);
+        }
+
+        @Override
+        protected final int getNextGatherOffset(long rootIndex) {
+            return (int) values.get(rootIndex);
+        }
+
+        @Override
+        protected final void setNextGatherOffset(long rootIndex, int offset) {
+            values.set(rootIndex, offset);
+        }
+
+        @Override
+        protected final boolean betterThan(long lhs, long rhs) {
+            return getOrder().reverseMul() * Long.compare(values.get(lhs), values.get(rhs)) < 0;
+        }
+
+        @Override
+        protected final void swap(long lhs, long rhs) {
+            long tmp = values.get(lhs);
+            values.set(lhs, values.get(rhs));
+            values.set(rhs, tmp);
+        }
+
+        // Type-specific methods
+
+        /**
+         * Collects a value into a bucket.
+         */
+        public void collect(long bucket, long value) {
+            values.set(0, value);
+            super.collect(bucket);
+        }
+
+        @Override
+        public List<Long> getValues(long bucket) {
+            var bounds = getBucketValuesBounds(bucket);
+            var size = bounds.v2() - bounds.v1();
+
+            if (size == 0) {
+                return emptyList();
+            }
+
+            var list = new ArrayList<Long>();
+
+            for (long i = bounds.v1(); i < bounds.v2(); i++) {
+                list.add(values.get(i));
+            }
+
+            list.sort(getOrder().wrap(Comparator.<Long>naturalOrder()));
+
+            return list;
+        }
```

Comment:
You might be wondering, "Nik, you go to all this trouble to heapify the data, why are you throwing your hands up and cramming it into a boxed list in aggregations and then sorting it?!" Well, the answer is "aggs never really had an efficient way to send this data, so it didn't matter much. And the "build" phase was always slower in aggs. But we, in ESQL-land, are better than aggs. In an ideal world we'd read the heapified data and dump it into a `BlockBuilder`. That's fun code to write. I just never did it.

Reply:
Used an array here to sort the elements, as to avoid boxing. Didn't use the heaps at all, as (I think) the complexity is quite similar if not identical